src/core/text-buffer.cc

```diff
@@ -378,6 +383,138 @@ struct TextBuffer::Layer {
+    size_t query_index = 0;
+    Point position;
+    Point current_word_start;
+    u16string current_word;
+
+    std::transform(query.begin(), query.end(), query.begin(), std::towlower);
+
+    // First, find the start position of all words matching the given
+    // subsequence.
+    std::unordered_map<u16string, vector<Point>> substring_matches;
+
+    for_each_chunk_in_range(Point(), extent(), [&] (TextSlice chunk) -> bool {
+      for (uint16_t c : chunk) {
+        bool is_word_character =
+          std::iswalnum(c) ||
+          std::find(extra_word_characters.begin(), extra_word_characters.end(), c) != extra_word_characters.end();
+
+        if (is_word_character) {
+          if (current_word.empty()) current_word_start = position;
+          current_word += c;
+          if (query_index < query.size() && towlower(c) == query[query_index]) {
+            query_index++;
+          }
+        } else {
+          if (!current_word.empty()) {
+            if (query_index == query.size()) {
+              substring_matches[current_word].push_back(current_word_start);
+            }
+            query_index = 0;
+            current_word.clear();
+          }
+        }
+
+        if (c == '\n') {
+          position.row++;
+          position.column = 0;
+        } else {
+          position.column++;
+        }
+      }
+
+      return false;
+    });
+
+    if (!current_word.empty() && query_index == query.size()) {
+      substring_matches[current_word].push_back(current_word_start);
+    }
+
+    // Next, calculate a score for each word indicating the quality of the
+    // match against the query.
+
+    static const unsigned consecutive_bonus = 5;
+    static const unsigned subword_start_bonus = 10;
+    static const unsigned mismatch_penalty = 1;
+    static const unsigned leading_mismatch_penalty = 3;
+
+    vector<SubsequenceMatch> matches;
+
+    for (auto entry : substring_matches) {
+      const u16string &word = entry.first;
+      const vector<Point> &start_positions = entry.second;
+
+      vector<SubsequenceMatchVariant> match_variants {{}};
+
+      for (size_t i = 0; i < word.size(); i++) {
+        uint16_t c = towlower(word[i]);
+
+        for (size_t j = 0, size = match_variants.size(); j < size; j++) {
+          SubsequenceMatchVariant &match_variant = match_variants[j];
+
+          if (match_variant.query_index < query.size()) {
+            // If the current word character matches the next character of
+            // the query for this match variant, create a *new* match variant
+            // that consumes the matching character.
+            if (c == query[match_variant.query_index]) {
+              SubsequenceMatchVariant new_match = match_variant;
+              new_match.query_index++;
+
+              if (i == 0 ||
+                  !std::iswalnum(word[i - 1]) ||
+                  (std::iswlower(word[i - 1]) && std::iswupper(word[i]))) {
```

Comment:
It seems like if not for this rule, there could be a function that takes any list of match indices and computes the score, but with this rule, I guess you also need to know whether each index corresponds to an upper or lower-case letter.